src/main/resources/xml/schema/2020-1/nordic2020-1.opf.sch

```diff
@@ -49,22 +49,15 @@
                 attribute</assert>
-            <assert test="count(dc:identifier) = 1 and matches(dc:identifier/text(),'^[A-Za-z0-9].*$')">[opf3a] The identifier ("<value-of select="dc:identifier/text()"/>") must start with a upper- or
-                lower-case letter (A-Z or a-z), or a digit (0-9).</assert>
-            <assert test="count(dc:identifier) = 1 and matches(dc:identifier/text(),'^.*[A-Za-z0-9]$')">[opf3a] The identifier ("<value-of select="dc:identifier/text()"/>") must end with a upper- or
-                lower-case letter (A-Z or a-z), or a digit (0-9).</assert>
-            <assert test="count(dc:identifier) = 1 and matches(dc:identifier/text(),'^[A-Za-z0-9_-]*$')">[opf3a] The identifier ("<value-of select="dc:identifier/text()"/>") must only contain upper-
-                or lower-case letters (A-Z or a-z), digits (0-9), dashes (-) and underscores (_).</assert>
-
-            <assert test="count(dc:title[not(@refines)]) = 1">[opf3b] exactly one dc:title <value-of select="if (dc:title[@refines]) then '(without a &quot;refines&quot; attribute)' else ''"/> must be
-                present in the package document.</assert>
-            <assert test="string-length(normalize-space(dc:title[not(@refines)]/text()))">[opf3b] the dc:title <value-of
-                    select="if (dc:title[@refines]) then '(without a &quot;refines&quot; attribute)' else ''"/> must not be empty.</assert>
+
+            <assert test="count(dc:title[@id='maintitle']) = 1 or count(dc:title) = 1">[opf3b] exactly one dc:title must be present in the package document when no main title is specified.</assert>
+            <assert test="count(dc:title[@id='maintitle']) = 0 or count(dc:title) > 1">[opf3b] more than one dc:title must be present in the package document when a main title is specified.</assert>
+            <assert test="string-length(normalize-space(dc:title[1]/text()))">[opf3b] the dc:title must not be empty.</assert>
 
             <assert test="count(dc:language[not(@refines)]) = 1">[opf3c] exactly one dc:language <value-of select="if (dc:language[@refines]) then '(without a &quot;refines&quot; attribute)' else ''"
                 /> must be present in the package document.</assert>
-            <assert test="count(dc:language[not(@refines)]) = 1 and matches(dc:language[not(@refines)]/text(), '^[a-z][a-z](-[A-Z][A-Z])?$')">[opf3c] the language code ("<value-of
-                    select="dc:language[not(@refines)]/text()"/>") must be either a "two-letter lower case" code or a "two-letter lower case + hyphen + two-letter upper case" code.</assert>
+            <assert test="count(dc:language[not(@refines)]) = 1 and matches(dc:language[not(@refines)]/text(), '^[a-z]{2}(-[A-Z]{2,3})?$')">[opf3c] the language code ("<value-of
+                    select="dc:language[not(@refines)]/text()"/>") must be either a "two-letter lower case" code or a "two-letter lower case + hyphen + two- or three-letter upper case" code.</assert>
```

Comment:
Yeah, it could be that someone asks for such a language code in editing instructions I guess, and it would be annoying if it turned out to be invalid.

How about:

- two or three ascii letters (a-z)
- zero or more sets of a hyphen, followed by one or more alphanumeric characters

So: `^[a-z]{2,3}(-[A-Za-z0-9]+)*$`

That would make sure that if there are no hyphens (most or all cases), there must be two or three lower case letters. And if you include one or more hyphens, it's relatively relaxed about the values following those hyphens.

What do you think?

Reply:
Sounds great, updated rule and text.